### PR TITLE
DISPATCH-1439 - Added two attributes to connection and one to router.

### PR DESCRIPTION
**Pull request by ganeshmurthy**
1. connection uptime - The number of seconds the connection has been up
2. connection last delivery seconds - The number of seconds since a delivery was sent on a connection
3. router uptime - the number of seconds since the router started

### CODE REVIEW

File: docs/man/qdstat.8.adoc [hunk 1]
```diff
@@ -73,6 +73,13 @@ The authentication method and user ID of the connection's authenticated user.
 tenant::
 If the connection is to a listener using multi-tenancy, this column displays the tenant namespace for the connection.
 
+last dlv::
+Time since the last delivery showed up on this connection. Displayed in the days:hours:minutes:seconds format.
```
Comment (bhardesty) [inline docs/man/qdstat.8.adoc]: Maybe "arrived" instead of "showed up"?

File: src/router_core/agent_connection.c [hunk 2]
```diff
@@ -240,6 +245,17 @@ static void qdr_connection_insert_column_CT(qdr_core_t *core, qdr_connection_t *
...
 
+    case QDR_CONNECTION_UPTIME_SECONDS:
+        qd_compose_insert_uint(body, core->uptime_ticks - conn->conn_uptime);
+        break;
+
+    case QDR_CONNECTION_LAST_DLV_SECONDS:
+        if (conn->is_delivery_sent)
+            qd_compose_insert_uint(body, core->uptime_ticks - conn->last_delivery_time);
+        else
+            qd_compose_insert_string(body, "-");
```
Comment (ted-ross) [inline src/router_core/agent_connection.c]: Adding hyphens to fields is the job of qdstat.  It should not be done here (I don't believe there is a precedent for it).  A better option is to insert a null value if there is no information to send.

File: src/router_core/agent_router.c [hunk 3]
```diff
@@ -233,6 +235,13 @@ static void qdr_agent_write_column_CT(qd_composed_field_t *body, int col, qdr_co
         qd_compose_insert_uint(body, core->links_blocked);
         break;
 
+    case QDR_ROUTER_UPTIME_SECONDS:
+        if (core->uptime_ticks== 0)
+            qd_compose_insert_uint(body, 0);
```
Comment (ted-ross) [inline src/router_core/agent_router.c]: This if-then-else doesn't seem to be any different from simply calling qs_compose_insert_uint.

File: src/router_core/router_core_private.h [hunk 4]
```diff
@@ -659,6 +659,9 @@ struct qdr_connection_t {
     qdr_conn_admin_status_t     admin_status;
     qdr_error_t                *error;
     bool                        closed; // This bit is used in the case where a client is trying to force close this connection.
+    uint32_t                    conn_uptime; // The number of seconds that this connection has been up and running.
+    uint32_t                    last_delivery_time; // The number of seconds since the last delivery arrived on this connection.
```
Comment (ted-ross) [inline src/router_core/router_core_private.h]: Both of these comments are incorrect.  These are timestamps, not durations.

File: src/router_core/router_core_private.h [hunk 5]
```diff
@@ -659,6 +659,9 @@ struct qdr_connection_t {
...
+    bool                        is_delivery_sent;
```
Comment (ted-ross) [inline src/router_core/router_core_private.h]: Couldn't this boolean be avoided by saying that last_delivery_time == 0 if no deliveries have been sent?

File: python/qpid_dispatch_internal/tools/display.py [hunk 6]
```diff
@@ -45,7 +45,14 @@ def Commas(value):
...
 def TimeLong(value):
-  return strftime("%c", gmtime(value / 1000000000))
+  day = value // (24 * 3600)
+  time = value % (24 * 3600)
+  hour = time // 3600
+  time %= 3600
+  minutes = time // 60
+  time %= 60
+  seconds = time
+  return "%03d:%02d:%02d:%02d" % (day, hour, minutes, seconds)
```
Comment (ted-ross) [inline python/qpid_dispatch_internal/tools/display.py]: I'm curious as to why this was done.  Does the strftime function no format the data properly?

Comment (ganeshmurthy) [reply python/qpid_dispatch_internal/tools/display.py]: I wanted qdstat to display in days:hours:minutes:seconds (we have a 3 digit limit on the days). I assumed that the router (or a connection) will not be up for more than 999 days.

That being said, strftime does not directly provide what I want. It rolls the the day into the year after 365 days. Then I need to use the year to calculate the days which becomes difficult if a leap year is involved. I could do the following and display the output in years:days:hours:minutes:seconds: 

value=150000000
year_from_epoch = strftime("%Y", gmtime(value))
year=int(year_from_epoch) - 1970
print ("Output=%s:%s" % (str(year), strftime("%j:%X", gmtime(value))))

[gmurthy@localhost test]$ python strtime.py 
Output=4:276:02:40:00
[gmurthy@localhost test]$ 

With all this in mind, I decided to write a simple custom function myself.

File: src/router_core/router_core_private.h (hunk 5 above)
Comment (ganeshmurthy) [reply src/router_core/router_core_private.h]: Yes, agreed

File: src/router_core/agent_connection.c (hunk 2 above)
Comment (ganeshmurthy) [reply src/router_core/agent_connection.c]: Changed this to insert null instead of "-"

File: src/router_core/router_core_private.h (hunk 4 above)
Comment (ganeshmurthy) [reply src/router_core/router_core_private.h]: Fixed comment

**Pull request closed without merging**